wire-ios/Wire-iOS/Sources/UserInterface/Calling/Analytics/CallEndedAnalyticsController.swift

```diff
@@ -0,0 +1,288 @@
+        ) { [weak self] notification in
+            self?.handleToggleVideoNotification(notification)
+        }
+    }
+
+    private func handleIncomingCall(
+        _ conversation: ZMConversation,
+        _ isVideoCall: Bool
+    ) {
+        if let eventInfo {
+            logger.error("handleIncomingCall: expected eventInfo to be nil, but is: \(eventInfo)")
+        }
+
+        eventInfo = .init(
+            callStart: currentDateProvider.now,
+            conversationID: conversation.avsIdentifier,
+            conversationType: conversation.conversationType == .group ? .group : .oneOnOne,
+            isVideoCall: isVideoCall
+        )
+    }
+
+    private func handleOutgoingCall(
+        _ conversation: ZMConversation,
+        _ isVideoCall: Bool
+    ) {
+        if let eventInfo {
+            logger.error("handleOutgoingCall: expected eventInfo to be nil, but is: \(eventInfo)")
+        }
+
+        eventInfo = .init(
+            callDirection: .outgoing,
+            callStart: currentDateProvider.now,
+            conversationType: conversation.conversationType == .group ? .group : .oneOnOne,
+            isVideoCall: isVideoCall
+        )
+    }
+
+    private func handleCallEstablished(_ conversation: ZMConversation) {
+        if eventInfo == nil {
+            logger.error("handleCallEstablished: expected eventInfo to be non-nil")
+            eventInfo = .init(callStart: currentDateProvider.now)
+        }
+
+        callParticipantObsererToken = CallCenter.addCallParticipantObserver(
+            observer: self,
+            for: conversation,
+            contextProvider: contextProvider
+        )
+    }
+
+    private func handleToggleVideoNotification(_ notification: Notification) {
+        guard
+            let conversationID = notification.userInfo?[CallCenter.conversationIDUserInfoKey] as? AVSIdentifier,
+            conversationID == eventInfo?.conversationID
+        else { return }
+
+        eventInfo?.hasAVSwitchToggled = true
+
+        // unsubscribe
+        toggleVideoObservationToken.map { notificationCenter.removeObserver($0) }
+        toggleVideoObservationToken = nil
+    }
+
+    private func handleCallTerminating(
+        _ conversation: ZMConversation,
+        _ reason: CallClosedReason
+    ) {
+        if eventInfo == nil {
+            logger.error("handleCallTerminating: expected eventInfo to be non-nil")
+            eventInfo = .init(callStart: currentDateProvider.now)
+        }
+
+        callParticipantObsererToken = nil
+
+        guard let eventInfo, let context = conversation.managedObjectContext else { return }
+
+        let (
+            isTeamMember,
+            conversationSize,
+            conversationGuestsTeam,
+            conversationGuestsNonTeam,
+            conversationServices
+        ) = context.performAndWait {
+
+            let isTeamMember = conversation.participants
+                .first { $0.isSelfUser }
+                .map(\.hasTeam) ?? false
```

Comment:
I think it would be more efficient to fetch the self user directly rather than finding them in the participant list. What do you think?

Reply:
Not sure, but I changed it.